### PR TITLE
Parse and apply table classes given alongside caption text

### LINKED ISSUE #4036: Support usage of classes given as Pandoc attributes (provided alongside table captions)
It would be good to migrate any provided classes (particularly as they may be Bootstrap classes) to the HTML table `<table>` element when they are present as Pandoc attributes (next to caption input). Here is an example of valid input:

```
| Default | Left | Right | Center |
|---------|:-----|------:|:------:|
| 12      | 12   |    12 |   12   |
| 123     | 123  |   123 |  123   |
| 1       | 1    |     1 |   1    |

: {#tbl-asdf .striped .hover .small tbl-colwidths="[15, 15, 35, 35]"}
```

This is a captionless table (though attributes can still be provided) and the ID, three classes, and the `tbl-colwidths` attribute (already supported) are provided. Note that these must be provided in that general order (id, classes, attributes). Also note that we should be flexible in allowing either `.table-striped` or the shorter `.striped` since we should resolve to the correct Bootstrap class names internally.

A final consideration is that the allowed classes should apply to the entire table (like the ones in the above example) since we will unconditionally apply them to the `<table>` element (and we don't have a sensible means to applying classes elsewhere in a table).

### PR DESCRIPTION
**Pull request by rich-iannone**
This allows for any table classes (should commonly be Bootstrap table classes) given as attributes next to a table caption to be parsed and inserted into the `<table>` element as classes. Any classes missing the necessary `table-` prefix will be handled so long as they are in the list of recognized Bootstrap 5 classes (those that apply to the entire table, not row, cell, or otherwise).

This is implemented as the `"pre-tableClasses"` Lua filter, placed in `"quartoPre"` *after* the `"pre-tableColwidth"` filter (which does parsing of the same attributes). The following Quarto document now produces a table with the Bootstrap classes applied:

````
---
format: html
---

| Default | Left | Right | Center |
|---------|:-----|------:|:------:|
| 12      | 12   |    12 |   12   |
| 123     | 123  |   123 |  123   |
| 1       | 1    |     1 |   1    |

: Here is the caption {.striped .hover .sm tbl-colwidths="[15, 15, 35, 35]"}
````

Here is the rendered document:

table-classes

Fixes: https://github.com/quarto-dev/quarto-cli/issues/4036

### CODE REVIEW

File: src/resources/filters/quarto-pre/table-classes.lua [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		if v == x then return true end
+	end
+	return false
+end
+
+-- propagate table classes to the table's <table> element
+function tableTableTag() 
+  return {
+    Table = function(el)
+      if tcontains(el.attr.classes) then
```
Comment (cscheid) [inline src/resources/filters/quarto-pre/table-classes.lua:16]: Are we missing a parameter here?

Relatedly, if you're using `tcontains` here, why do we define `contains` just before that?

File: src/resources/filters/quarto-pre/table-classes.lua [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      
+      -- if there is no caption then return tbl unchanged
+      if tbl.caption.long == nil or #tbl.caption.long < 1 then
+        return tbl
+      end
+
+      -- determine if we have any supplied classes, these should always begin with a `.` and
+      -- consist of alphanumeric characters
+      local caption =  tbl.caption.long[#tbl.caption.long]
+      local caption_str = pandoc.utils.stringify(caption)
```
Comment (cscheid) [inline src/resources/filters/quarto-pre/table-classes.lua:47]: We should be using `parseTableCaption` here, like it's done in table-colwidth.lua

File: src/resources/filters/quarto-pre/table-classes.lua [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        "primary", "secondary", "success", "danger", "warning", "info", "light", "dark",
+        "striped", "hover", "active", "bordered", "borderless", "sm",
+        "responsive", "responsive-sm", "responsive-md", "responsive-lg", "responsive-xl", "responsive-xxl"
+      }
+
+      -- strip off leading `.` chars from each class in `css_classes` and prefix each
+      -- item that is recognized as a Bootstrap class with `table-` 
+      for k, v in pairs(css_classes) do
+        css_classes[k] = string.sub(v, 2)
+        if contains(table_bootstrap_nm, css_classes[k]) then
```
Comment (cscheid) [inline src/resources/filters/quarto-pre/table-classes.lua:71]: we have `tcontains` in our utils.

**Review by cscheid — CHANGES_REQUESTED** [review]
We should also add test cases.

File: src/resources/filters/quarto-pre/table-classes.lua [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      -- generate a table containing recognized Bootstrap table classes
+      local table_bootstrap_nm = {
+        "primary", "secondary", "success", "danger", "warning", "info", "light", "dark",
+        "striped", "hover", "active", "bordered", "borderless", "sm",
+        "responsive", "responsive-sm", "responsive-md", "responsive-lg", "responsive-xl", "responsive-xxl"
+      }
+
+      -- strip off leading `.` chars from each class in `css_classes` and prefix each
+      -- item that is recognized as a Bootstrap class with `table-` 
+      for k, v in pairs(css_classes) do
```
Comment (allenmanning) [inline src/resources/filters/quarto-pre/table-classes.lua:69]: Looking great Rich!  May I suggest extracting pure functions out to unit tests?

**Review by cscheid — APPROVED** [review]
lgtm assuming test suite passes.

**Pull request merged**